### DISCUSSION TITLE
Mouse in PC basic?

### DISCUSSION THREAD

**Discussion by smithdoor** [robhagemans/pcbasic#174] [General]
Any word on using a Mouse in PC basic or how to use a mouse in PC Basic?

Dave

**Comment by robhagemans** [comment]
GW-BASIC pre-dated the common use of mice on PC's so never had support for it. In PC-BASIC, the mouse can be used to emulate the light pen (for which GW-BASIC *did* have support) e.g. https://robhagemans.github.io/pcbasic/doc/2.0/#PEN

**Comment by smithdoor** [comment]
Thank youI had 2.02 but just download the new 2.04 and try the mouse using the pen.Works great Please keep up the good work.Since 2020 I have ben on chemo and was not up to programing. Now getting back into programing. Dave Sent from Mail for Windows From: Rob HagemansSent: Thursday, January 13, 2022 11:26 PMTo: robhagemans/pcbasicCc: smithdoor; AuthorSubject: Re: [robhagemans/pcbasic] Mouse in PC basic? (Discussion #174) GW-BASIC pre-dated the common use of mice on PC's so never had support for it. In PC-BASIC, the mouse can be used to emulate the light pen (for which GW-BASIC did have support) e.g. https://robhagemans.github.io/pcbasic/doc/2.0/#PEN—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you authored the thread.Message ID: ***@***.***>

**Comment by smithdoor** [comment]
> Any word on using a Mouse in PC basic or how to use a mouse in PC Basic?
> 
> Dave
I try using the mouse in 2.0.4 it works great.

Thank for great work 
Below is a simple program for mouse used as a pen.

Dave 

10 ' *** Mouse used in PC Basic 2.0.4
20 COLOR 14,1
30 CLS
40 PEN ON
50 FOR I=1 TO 5000 ' *** Typically INKEY$ is used
60 LOCATE 3,6:PRINT "Left click mouse to end";
70 LOCATE 5,1:PRINT PEN(0),PEN(1),PEN(2),PEN(3),PEN(4),PEN(5),PEN(6),PEN(7),PEN(8),PEN(9)
80 LOCATE 8,10:COLOR 15,1:PRINT "horizontal{PEN(4) ";PEN(4);" * vertical {PEN(5)} ";PEN(5):COLOR 14,0
90 IF PEN(3)=-1  THEN COLOR 15,1:LOCATE 14,3: PRINT "locate of mouse Left click {PEN(3)}= ";PEN(3); "horizontal = ";PEN(4);" X  vertical = ";PEN(5):LOCATE 18,12:PRINT "         end":END
100 NEXT'  *** Typically GOTO (INKEY$) is used
110 PEN OFF


**Comment by smithdoor** [comment]
Here is simple program I use to put into a dice game using the keyboard now use the mouse
Without a modifying the game. 

10 ' *** Mouse used in PC Basic 2.0.4
20 ' data for location to fit in row 10 to 24 and column 1 to 36
30 '  ***    XPEN25= ROW  *** XPEN24 =  COLUMN  
40 '
50 CLS
60 COLOR 14,2
62 '
63 PEN ON
64 for IBGR= 11 to 21
66 LOCATE IBGR,2:PRINT STRING$(28,32);
68 next IBGR
69 LOCATE 22,2:PRINT STRING$(24,32);STRING$(3,219);" ";
70 LOCATE 23,2:PRINT STRING$(4,32);"Click to pickup cup ";STRING$(3,219);" ";
72 LOCATE 24,2:PRINT STRING$(28,32); 
73 '
74 if pen(3)=-1 then 135
78 goto 74
80 ' ***
130 '
135 ' *** start dice shake
140  AINKEY$=INKEY$
160 '
165 ' *** Pen data for display for mouse
170 XPEN25= (INT((PEN(5) /16) +1.3))'   *** used for pen die  ROW12+6
180 XPEN24= (INT((PEN(4) /8.2)-1)) '   *** COLUMN
190 '
200 ' ***
210 ' *** Check for over mouse/pen 
220 IF XPEN24<2 THEN XPEN24=2      ' ***  COLUMN
230 IF XPEN24>30 THEN XPEN24=30
240 IF XPEN25<12  THEN XPEN25=12     ' *** ROW
250 IF XPEN25>23 THEN XPEN25=23
260 '
270 color 7,1
275 LOCATE 1,1: PRINT "ROW 25 = ";XPEN25;"   *** COLUMN 24=";XPEN24,dxp,s$' *** used for testing
280 LOCATE 2,1: PRINT "ROW 25 = "PEN(5);"   *** COLUMN 24=";PEN(4);' *** used for testing
290 ' 
300 FOR IBGR = 11 TO 24
310 IF IBGR=XPEN25+1 THEN LOCATE XPEN25+1,2:COLOR 14,2:PRINT STRING$(XPEN24-2,32);STRING$(3,219);STRING$(30-XPEN24,32);:GOTO 340
320 IF IBGR=XPEN25 THEN LOCATE XPEN25,2:COLOR 14,2:PRINT STRING$(XPEN24-2,32);STRING$(3,219);STRING$(30-XPEN24,32);:GOTO 340
330 LOCATE IBGR,2:PRINT STRING$(32,32);
334 '
335 If  PEN(5) <10  THEN 400 ' *** Dice are thrown
336 IF XPEN24 > 15 and dxp>0 THEN SOUND 60, .05:SOUND 45, .07:SOUND 40, .05:SOUND 37, .07:SOUND 45, .07:SOUND 40, .05:SOUND 37, .07
338 IF XPEN24 <12 and dxp<1  THEN SOUND 120, .05:SOUND 70, .07:SOUND 80, .05:SOUND 40, .05:SOUND 70, .07:SOUND 50, .07:SOUND 120, .05:SOUND 70, .07:SOUND 80, .05:SOUND 40, .05
340 NEXT IBGR
350 '  
360 '
370 IF AINKEY$ = CHR$(27) THEN PEN OFF:COLOR 14,1:END
380 GOTO 140 ' *** GOTO (INKEY$) 
390 '
400 ' next step
410 color 14,1
420 locate 1,1:print "            ********* Dice is on the table
430 color 15,1

**Comment by smithdoor** [comment]
The mouse is working great.
It close to programing a mouse (RS232 port) back earlier 1980's. 
It is great job for the Robhagemans and team 

Thank you 

Dave
